### PR TITLE
DAOS-3919 test: Re-strucutring iorsmall and mdtest

### PR DESCRIPTION
**Pull request by saurabhtandan**
Removing the use of !mux from iorsmall and mdtestsmall
tests in order to reduce the run time.

Signed-off-by: Saurabh Tandan <saurabh.tandan@intel.com>

### CODE REVIEW

File: src/tests/ftest/io/mdtest_small.py [hunk 1]
```diff
@@ -44,6 +44,62 @@ def test_mdtest_small(self):
...
+        number_of_files_dirs = self.params.get("number_of_files_dirs",
+                                               "/run/mdtest/*")
+
+        # Running mdtest for different variants
+        for flag in mdtest_flags:
+            self.mdtest_cmd.flags.update(flag)
+            if self.mdtest_cmd.flags.value == ' ':
+                for api in mdtest_api:
+                    self.mdtest_cmd.api.update(api)
+                    if self.mdtest_cmd.api.value == 'DFS':
```
Comment (Puspalogan) [inline src/tests/ftest/io/mdtest_small.py]: Would recommend similar approach to map the variants as comma separated to avoid multiple if/else conditions.

File: src/tests/ftest/io/ior_small.py [hunk 2]
```diff
@@ -48,4 +48,32 @@ def test_ior_small(self):
...
+        obj_class = self.params.get("obj_class", '/run/ior/iorflags/*')
+
+        for oclass in obj_class:
+            self.ior_cmd.daos_oclass.update(oclass)
+            for api in apis:
+                self.ior_cmd.api.update(api)
+                self.ior_cmd.flags.update(flags[0])
+                for ts in transfer_size:
+                    self.ior_cmd.transfer_size.update(ts)
+                    if ts == '256B':
```
Comment (Puspalogan) [inline src/tests/ftest/io/ior_small.py]: There is a mapping between transfer_size and block_size. To avoid the if condition, wouldn't it be easier to read comma separated transfer_size,blocksize in the yaml and split it here ? That way we wouldn't need if condition and more variants can be added in future without having to add additional elsif conditions.

File: src/tests/ftest/io/ior_small.yaml [hunk 3]
```diff
@@ -11,7 +11,7 @@ hosts:
     - client-F
     - client-G
     - client-H
-timeout: 1800
+timeout: 3000
```
Comment (Puspalogan) [inline src/tests/ftest/io/ior_small.yaml]: Shouldn't the timeout be lower than 1800 if we're avoiding server start and disk formatting for each test variant ?

Comment (rpadma2) [reply src/tests/ftest/io/ior_small.yaml]: 3000 seconds seems to be a long time. Is it not?

File: src/tests/ftest/io/mdtest_small.py [hunk 4]
```diff
@@ -44,6 +44,62 @@ def test_mdtest_small(self):
...
+                        self.mdtest_cmd.write_bytes.update(read_write[0])
+                        self.mdtest_cmd.read_bytes.update(read_write[0])
+                        self.mdtest_cmd.branching_factor.update(
+                            branch_factor[1])
+                        self.mdtest_cmd.items.update(num_of_items)
+                        self.mdtest_cmd.depth.update(dir_depth[2])
+
+                    self.execute_mdtest()
+                    self.mdtest_cmd.num_of_files_dirs.update(None)
+                    self.mdtest_cmd.items.update(None)
```
Comment (rpadma2) [inline src/tests/ftest/io/mdtest_small.py]: Can we add some comments to this big if # else and for loops? It might be difficult for someone to understand what is going on which this piece of code without putting some debug statements.

File: src/tests/ftest/io/mdtest_small.yaml [hunk 5]
```diff
@@ -7,7 +7,7 @@ hosts:
   test_clients:
     - client-E
     - client-F
-timeout: 800
+timeout: 1500
```
Comment (rpadma2) [inline src/tests/ftest/io/mdtest_small.yaml]: Why increasing timeout values? Is something going which warrants big increase (eg: scm_size??)?

File: src/tests/ftest/io/ior_small.yaml (hunk 3 above)
Comment (saurabhtandan) [reply src/tests/ftest/io/ior_small.yaml]: This I kept high for my initial runs of CI to make sure test finishes in CI. SO was just to be on safe side. It actually took considerably less than what I thought it would. So changing it to 1800 now as in one of CI runs it finished in 20 mints.
Also previously the timeouts were for each test variants. Now all different test variants are being run as part of single test. So having higher timeout value. Will see how long these tests consistently and we can change these values in future.

File: src/tests/ftest/io/mdtest_small.yaml (hunk 5 above)
Comment (saurabhtandan) [reply src/tests/ftest/io/mdtest_small.yaml]: Same as above. Not changing it yet.

File: src/tests/ftest/io/ior_small.py [hunk 6]
```diff
@@ -48,4 +48,30 @@ def test_ior_small(self):
...
+        transfer_block_size = self.params.get("transfer_block_size",
+                                              '/run/ior/iorflags/*')
+        obj_class = self.params.get("obj_class", '/run/ior/iorflags/*')
+
+        # run tests for different variants
+        for oclass in obj_class:
+            self.ior_cmd.daos_oclass.update(oclass)
+            for api in apis:
+                self.ior_cmd.api.update(api)
+                self.ior_cmd.flags.update(flags[0])
```
Comment (Puspalogan) [inline src/tests/ftest/io/ior_small.py]: If you're going to make another commit, you can move line #62 before #58 and mention in the comment that it's for ssf.

File: src/tests/ftest/io/ior_small.py [hunk 7]
```diff
@@ -48,4 +48,30 @@ def test_ior_small(self):
...
+        for oclass in obj_class:
+            self.ior_cmd.daos_oclass.update(oclass)
+            for api in apis:
+                self.ior_cmd.api.update(api)
+                for test in transfer_block_size:
+                    # update transfer and block size
+                    self.ior_cmd.transfer_size.update(test[0])
+                    self.ior_cmd.block_size.update(test[1])
+                    # run ior
+                    self.run_ior_with_pool()
```
Comment (phender) [inline src/tests/ftest/io/ior_small.py]: In the case of the `256B` transfer size, it doesn't look like the user is actually notified that the case is skipped.  This function now returns `"Skipping the case for transfer_size=256B"`, but it is not displayed.  One possible solution would be to have the `run_ior_with_pool` method log/print the message and just return `None`.

File: src/tests/ftest/io/ior_small.py [hunk 8]
```diff
@@ -48,4 +48,30 @@ def test_ior_small(self):
...
+
+        # Running a variant for ior fpp
+        self.ior_cmd.flags.update(flags[1])
```
Comment (phender) [inline src/tests/ftest/io/ior_small.py]: How concerned are we that the test should not produce python errors if the yaml file inputs changed?  This section for running ior fpp is highly dependent upon specific indices of various yaml inputs.  If someone want to shorten the run time of the test by commenting out a few items in the input lists this test would fail or even run with non-intended inputs (as it seems the ior fpp inputs are specific).

File: src/tests/ftest/io/ior_small.py (hunk 7 above)
Comment (saurabhtandan) [reply src/tests/ftest/io/ior_small.py]: Well is it okay if we take care of that in a different pr. Landing of this one can save decent amount of time. If you are fine we can go ahead with it. I will add the one liner that in my another PR which is not yet merged.

File: src/tests/ftest/io/ior_small.py (hunk 8 above)
Comment (saurabhtandan) [reply src/tests/ftest/io/ior_small.py]: I agree, but I think there are going to be such consequences of giving up with !mux for less time. If somebody wants to modify the yams they will simply need to comment these lines in the code too if they want to do some quick check.
